### PR TITLE
Add ability to grab and flash packages based on size

### PR DESCRIPTION
**Pull request by smcd253**

### CODE REVIEW

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 1]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
+              // save pointer to first chunk
+              az_span data = az_span_create(packet_ptr->nx_packet_prepend_ptr, 
+                                                  packet_ptr->nx_packet_length);
+
+              // calculate destination pointer
+              uint32_t total_downloaded_size = 0;
+              uint8_t* dest_ptr = (uint8_t*)(address) + total_downloaded_size;
+
+              // grab package size and round up to nearest 2KB (0x0800)
+              unsigned int package_size = (unsigned int)http_client.nx_web_http_client_total_receive_bytes;
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >unsigned int [](http://example.com/codeflow?start=14&length=12)

Please replalce by uint32_t #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 2]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
+              package_size += 0x0800 - package_size % 0x0800;
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >package_size += 0x0800 - package_size % 0x0800; [](http://example.com/codeflow?start=14&length=47)

Looks a little expensive for me, consider something like
```
if((package_size & 0x07FF) != 0x000)
{
  package_size = (package_size & 0xFFFFF800) + 0x0800;
}
``` #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 3]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
-            nx_status = nx_web_http_client_request_send(&http_client, DCF_WAIT_TIME);
-            // Download destination pointer, increase after every packet is stored
-            uint32_t total_downloaded_size = 0;
-            while(nx_status == NX_SUCCESS)
+          // send request
+          if((nx_status = nx_web_http_client_request_send(&http_client, 
+                                                          DCF_WAIT_TIME)) == NX_SUCCESS)
             {
+              // grab first chunk
               nx_status = nx_web_http_client_response_body_get(&http_client, &packet_ptr, 500);
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >nx_status [](http://example.com/codeflow?start=14&length=9)

Please test nx_status == NX_SUCCESS here. #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 4]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
+              package_size += 0x0800 - package_size % 0x0800;
 
-                // call store to flash
-                if((hal_status = internal_flash_write(
-                                      dest_ptr, az_span_ptr(data), az_span_size(data))) != HAL_OK)
+              // erase flash
+              if ((hal_status = internal_flash_erase((UCHAR*)address, package_size)) == HAL_OK)
+              {
+                if((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
                 {
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:232]: Please remove this test. #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 5]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
-                  nx_status = nx_packet_release(packet_ptr);
-                  break;
+                  // write first chunk to flash
+                  if((hal_status = internal_flash_write(
+                                        dest_ptr, az_span_ptr(data), az_span_size(data))) != HAL_OK)
+                  {
+                    nx_status = nx_packet_release(packet_ptr);
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >nx_status = nx_packet_release(packet_ptr); [](http://example.com/codeflow?start=20&length=42)

Considere move it outside of the `if` statement once it is inside of the `if` and the `else`. #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 6]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
-                                              packet_ptr->nx_packet_length);
 
-                // calculate destination pointer
-                uint8_t* dest_ptr = (uint8_t*)(address) + total_downloaded_size;
+              // save pointer to first chunk
+              az_span data = az_span_create(packet_ptr->nx_packet_prepend_ptr, 
+                                                  packet_ptr->nx_packet_length);
+
+              // calculate destination pointer
+              uint32_t total_downloaded_size = 0;
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >uint32_t total_downloaded_size = 0; [](http://example.com/codeflow?start=14&length=35)

Consider an inproviment here removing this `total_downloaded_size` variable. In this case, make 
```
uint8_t* dest_ptr = (uint8_t*)(address);
```
and add the data size directly to the `dest_ptr`.
```
dest_ptr += az_span_size(data);
```
Both outside and inside of the `while` loop.  #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 7]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
                 }
 
-                // update total package size
-                total_downloaded_size += az_span_size(data);
-
-                nx_status = nx_packet_release(packet_ptr);
+                // if there are more chunks to store, loop over them until done
+                while(nx_status == NX_SUCCESS)
+                {
+                  // grab next chunk
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >// grab next chunk [](http://example.com/codeflow?start=18&length=18)

Thanks for adding these comments.  #Closed

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c (hunk 3 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: || nx_status == NX_WEB_HTTP_GET_DONE

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 8]
```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
+
+              // grab package size and round up to nearest 2KB (0x0800)
+              unsigned int package_size = (unsigned int)http_client.nx_web_http_client_total_receive_bytes;
+              package_size += 0x0800 - package_size % 0x0800;
 
-                // call store to flash
-                if((hal_status = internal_flash_write(
-                                      dest_ptr, az_span_ptr(data), az_span_size(data))) != HAL_OK)
+              // erase flash
+              if ((hal_status = internal_flash_erase((UCHAR*)address, package_size)) == HAL_OK)
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:227]: >if ((hal_status = internal_flash_erase((UCHAR*)address, package_size)) == HAL_OK) [](http://example.com/codeflow?start=14&length=81)

You probably need to release `packet_ptr` if this statement return error.

**Review by mamokarz — DISMISSED** [review]
:clock1:

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c (hunk 8 above)
Comment (smcd253) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:227]: good call! think unit tests may be in order soon...

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c (hunk 2 above)
Comment (smcd253) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: oh wow yes much more elegant.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c (hunk 1 above)
Comment (smcd253) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: okay. going to change the flash code to take uint32_t as well

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 9]
```diff
@@ -174,70 +174,105 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
-                                    DCF_WAIT_TIME)) == NX_SUCCESS)
-          {
-            nx_status = nx_web_http_client_request_send(&http_client, DCF_WAIT_TIME);
-            // Download destination pointer, increase after every packet is stored
-            uint32_t total_downloaded_size = 0;
-            while(nx_status == NX_SUCCESS)
+          // send request
+          if((nx_status = nx_web_http_client_request_send(&http_client, 
+                                                          DCF_WAIT_TIME)) == NX_SUCCESS)
             {
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:211]: Please correct the indentation.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c (hunk 8 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:227]: The release is not in the right place, if `nx_web_http_client_response_body_get` failed, the `packet_ptr` shall not be released.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 10]
```diff
@@ -174,70 +174,105 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
-              if((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
+              if ((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
               {
+                // save pointer to first chunk
                 az_span data = az_span_create(packet_ptr->nx_packet_prepend_ptr, 
-                                              packet_ptr->nx_packet_length);
+                                                    packet_ptr->nx_packet_length);
 
                 // calculate destination pointer
+                uint32_t total_downloaded_size = 0;
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:217]: >uint32_t total_downloaded_size = 0; [](http://example.com/codeflow?start=16&length=35)

You can delete the variable `total_download_size`.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 11]
```diff
@@ -174,70 +174,105 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
+                                        dest_ptr, az_span_ptr(data), az_span_size(data))) == HAL_OK)
+                  {
+                    nx_status = nx_packet_release(packet_ptr);
+
+                    // if there are more chunks to store, loop over them until done
+                    while(nx_status == NX_SUCCESS)
+                    {
+                      // grab next chunk
+                      nx_status = nx_web_http_client_response_body_get(&http_client, &packet_ptr, 500);
+                        if((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >  if [](http://example.com/codeflow?start=22&length=4)

Indentention here as well. #Resolved

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c (hunk 8 above)
Comment (mamokarz) [reply STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c:227]: In the current code, you need to release in a `else` of the ifs in the lines 228 and 231.

File: STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c [hunk 12]
```diff
@@ -174,70 +174,104 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
...
-            uint32_t total_downloaded_size = 0;
-            while(nx_status == NX_SUCCESS)
+          // send request
+          if((nx_status = nx_web_http_client_request_send(&http_client, 
+                                                          DCF_WAIT_TIME)) == NX_SUCCESS)
             {
+              // grab first chunk
               nx_status = nx_web_http_client_response_body_get(&http_client, &packet_ptr, 500);
-              if((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
+              if ((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
```
Comment (mamokarz) [inline STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c]: >nx_status == NX_WEB_HTTP_GET_DONE [](http://example.com/codeflow?start=48&length=33)

If nx_status is `NX_WEB_HTTP_GET_DONE` and erase of write in flash failed, the function will return the wrong error code. So, on those situations, nx_status shall be converted to `NX_SUCCESS`.

**Pull request merged**